Add Multi-Fractional Non-Fungible Token Standard (#4675)
ethereum/EIPs: EIPS/eip-4675.md

---
eip: 4675

> title: Multi-Fractional Non-Fungible Token Standard
> description: Token standard to handle multiple NFTs in a fractionalized form

Reviewer: ```suggestion
title: Multi-Fractional Non-Fungible Tokens
description: Fractionalize multiple NFTs using a single contract
```

We try to avoid "standard" in titles and descriptions.

author: David Kim (@powerstream3604)
discussions-to: https://ethereum-magicians.org/t/eip-4675-multi-fractional-non-fungible-token-standard/8008
status: Draft
type: Standards Track
category: ERC
created: 2022-01-13
requires: 165, 721
---

## Abstract

> This standard outlines a smart contract interface eligible to represent any number of fractionalized non-fungible tokens. Existing projects utilizing standards like EIP-1633 conventionally deploy seperate ERC-20 compatible token standard to fractionalize the non-fungible token into ERC-20 tokens. In contrast, this ERC allows each token ID to represent a token type representing(fractionalizing) the non-fungible token.

Reviewer: The first mention of another EIP must be linked, use EIP instead of ERC, and I think you had the wrong word in here:

```suggestion
This standard outlines a smart contract interface eligible to represent any number of fractionalized non-fungible tokens. Existing projects utilizing standards like [EIP-1633](./eip-1633.md) conventionally deploy separate [EIP-20](./eip-20.md) compatible token contracts to fractionalize the non-fungible token into EIP-20 tokens. In contrast, this ERC allows each token ID to represent a token type representing(fractionalizing) the non-fungible token.
```

> This standard is approximate in terms of using `_id` for distinguishing token types. However, this ERC has a clear difference with ERC-1155 as each `_id` represents a distinct NFT.

Reviewer: ```suggestion
This standard is approximate in terms of using `_id` for distinguishing token types. However, this ERC has a clear difference with [EIP-1155](./eip-1155.md) as each `_id` represents a distinct NFT.
```

        @param interfaceID The interface identifier, as specified in ERC-165
        @dev Interface identification is specified in ERC-165. This function
        uses less than 30,000 gas.
        @return `true` if the contract implements `interfaceID` and
        `interfaceID` is not 0xffffffff, `false` otherwise
    */
    function supportsInterface(bytes4 interfaceID) external view returns (bool);
}
```

To receive Non-Fungible Token on `safe Transfer` the contract should include `onERC721Received()`.
Including `onERC721Received()` is needed to be compatible with Safe Transfer Rules.
```solidity
/**
    @notice Handle the receipt of an NFT
    @param _operator The address which called `safeTransferFrom` function
    @param _from The address which previously owned the token
    @param _tokenId The NFT identifier which is being transferred
    @param _data Additional data with no specified format
    @return `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`
*/
function onERC721Received(address _operator, address _from, uint256 _tokenId, bytes calldata _data) external pure returns (bytes4);
```

## Rationale

**Metadata**

> The `symbol()` & `name()` function was not included since the majority of users can just fetch it from the originating NFT contract. Also, copying the name & symbol every time when token gets added might place a lot of redundant bytecode on the Ethereum blockchain. 

Reviewer: ```suggestion
The `symbol()` & `name()` functions were not included since the majority of users can just fetch it from the originating NFT contract. Also, copying the name & symbol every time when token gets added might place a lot of redundant bytecode on the Ethereum blockchain. 
```

**Design**

Most of the decisions made around the design of this ERC were done to keep it as flexible for diverse token design & architecture.
These minimum requirement for this standard allows for each project to determine their own system for minting, governing, burning their MFNFT tokens depending on their programmable architecture.

## Backwards Compatibility

To make this standard compatible with existing standards, this standard `event` & `function` names are identical with ERC-20 token standard with some more `events` & `functions` to add token type dynamically.

Also, the sequence of parameter in use of `_id` for distinguishing token types in `functions` and `events` are very much similar to ERC-1155 Multi-Token Standard.

Since this standard is intended to interact with the EIP-721 Non-Fungible Token Standard, it is kept purposefully agnostic to extensions beyond the standard in order to allow specific projects to design their own token usage and scenario.

## Test Cases

Reference Implementation of MFNFT Token includes test cases written using hardhat. (Test coverage : 100%)

## Reference Implementation

> [MFNFT - Implementation](../assets/eip-4675/)

Reviewer: You might need to link to a specific file (like a README) to get the bot to accept your PR.

**Design**

Most of the decisions made around the design of this ERC were done to keep it as flexible for diverse token design & architecture.
These minimum requirement for this standard allows for each project to determine their own system for minting, governing, burning their MFNFT tokens depending on their programmable architecture.

## Backwards Compatibility

To make this standard compatible with existing standards, this standard `event` & `function` names are identical with ERC-20 token standard with some more `events` & `functions` to add token type dynamically.

Also, the sequence of parameter in use of `_id` for distinguishing token types in `functions` and `events` are very much similar to ERC-1155 Multi-Token Standard.

Since this standard is intended to interact with the EIP-721 Non-Fungible Token Standard, it is kept purposefully agnostic to extensions beyond the standard in order to allow specific projects to design their own token usage and scenario.

## Test Cases

Reference Implementation of MFNFT Token includes test cases written using hardhat. (Test coverage : 100%)

## Reference Implementation

> [MFNFT - Implementation](https://github.com/PowerStream3604/multi-fnft)

Reviewer: EIP-1 was recently updated to recommend against external links (Sorry!)

Would it be possible to include the reference implementation in the `assets` directory, as described [here](https://eips.ethereum.org/EIPS/eip-1#auxiliary-files)?

Author: I appreciate your update! @SamWilsn 
Deleted all external links and pointed it to ../assets

## Security Considerations

To fractionalize an already minted NFT, it is evident that ownership of NFT should be given to token contracts before fractionalization.
In the case of fractionalizing NFT, the token contract should thoroughly verify the ownership of NFT before fractionalizing it to prevent tokens from being a separate tokens with the NFT.

If an arbitrary account has the right to call `setParentNFT()` there might be a front-running issue. The caller of `setParentNFT()` might be different from the real NFT sender. 
To prevent this issue, implementors should just allow **admin** to call, or fractionalize and receive NFT in an atomic transaction similar to flash loan(swap).

## Copyright
Copyright and related rights waived via [CC0](../LICENSE.md).

> ## Citation

Reviewer 2: Please remove this section, it is added automatically on the website `eips.ethereum.org`.

Reviewer 2: @PowerStream3604, this comment is on the "Citation" section. The "Copyright" section is required.

Author: @lightclient , I updated the file as requested.
Sorry, I misunderstood it.

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
> 
> ## Citation
> 
> Please cite this document as:
> 
> [David Kim](mailto:helloyongjun3604@gmail.com) "EIP-4675: Multi-Fractional Non-Fungible Token"

Reviewer 3: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

---
eip: 4675
title: Multi-Fractional Non-Fungible Token Standard

> description: 

Reviewer: Should have one full short sentence here.

> author: David Kim(@powerstream3604)

Reviewer: ```suggestion
author: David Kim (@powerstream3604)
```

discussions-to: https://ethereum-magicians.org/t/eip-4675-multi-fractional-non-fungible-token-standard/8008
status: Draft
type: Standards Track
category: ERC
created: 2021-01-13

> requires: 165

Reviewer: Might also be worth mentioning ERC-721?

---

## Abstract
This standard outlines a smart contract interface eligible to represent any number of fractionalized non-fungible tokens. Existing projects utilizing standards like EIP-1633 conventionally deploy seperate ERC-20 compatible token standard to fractionalize the non-fungible token into ERC-20 tokens. In contrast, this ERC allows each token ID to represent a token type representing(fractionalizing) the non-fungible token.

> This standard is approximate in terms of using ``_id`` for distinguishing token types. However, this ERC has a clear difference with ERC-1155 as each ``_id`` represents a distinct NFT.

Author: This is already updated in the current version.
The current version uses a single back tick to denote code.

**Design**

Most of the decisions made around the design of this ERC were done to keep it as flexible for diverse token design & architecture.
These minimum requirement for this standard allows for each project to determine their own system for minting, governing, burning their MFNFT tokens depending on their programmable architecture.

## Backwards Compatibility

To make this standard compatible with existing standards, this standard ``event`` & ``function`` names are identical with ERC-20 token standard with some more ``events`` & ``functions`` to add token type dynamically.

Also, the sequence of parameter in use of ``_id`` for distinguishing token types in ``functions`` and ``events`` are very much similar to ERC-1155 Multi-Token Standard.

Since this standard is intended to interact with the EIP-721 Non-Fungible Token Standard, it is kept purposefully agnostic to extensions beyond the standard in order to allow specific projects to design their own token usage and scenario.

## Test Cases

Reference Implementation of MFNFT Token includes test cases written using hardhat. (Test coverage : 100%)

> ## References

Reviewer 4: There isn't a `References` section in [EIP-1](./eip-1.md) - please remove this section. In terms of the information, you can put it in a different section if you deem it necessary, but please use relative links, so instead of:
```
[ERC-20](https://eips.ethereum.org/EIPS/eip-20)
```
use:
```
[ERC-20](./erc-20.md)
```
